torch/distributed/_composable/contract.py

```diff
@@ -0,0 +1,145 @@
+import torch
+import torch.nn as nn
+
+from collections import OrderedDict
+from typing import Any, Callable, Dict, List, Optional
+
+
+# use state_slot as key for module.__dict__ to avoid coliding with other
+# properties.
+# TODO: since all composable distributed features can share the same slot.
+class _StateKey:
+
+    # implement operator < to avoid breaking dir()
+    def __lt__(self, other: Any) -> bool:
+        return True if isinstance(other, str) else id(self) < id(other)
+
+
+class _State:
+    pass
+
+
+state_key = _StateKey()
+
+
+def contract(func):
+    r"""
+    Decorate a function as a composable distributed API, where the first
+    argument of the function must be an :class:`nn.Module` instance. The
+    decorator verifies that the wrapped function does not modify parameter,
+    buffer or sub-module fully-qualified names (FQN).
+
+    When a function ``func`` is decorated by ``@contract``, a
+    ``.state(module: nn.Module)`` method will be installed to the decorated
+    function. Then you can retrieve and modify the state on a module by calling
+    ``func.state(module)``.
+
+    Example::
+        >>> import torch.nn as nn
+        >>>
+        >>> class MyModel(nn.Module):
+        >>>     def __init__(self):
+        >>>         super().__init__()
+        >>>         self.l1 = nn.Linear(10, 10)
+        >>>         self.l2 = nn.Linear(10, 10)
+        >>>
+        >>>     def forward(self, x):
+        >>>         return self.l2(self.l1(x))
+        >>>
+        >>> @contract
+        >>> def my_feature(module: nn.Module) -> nn.Module:
+        >>>     my_feature.state(module).some_state = "any value"
+        >>>     return module
+        >>>
+        >>> model = MyModel()
+        >>> my_feature(model.l1)
+        >>> assert my_feature.state(model.l1).some_state == "any value"
+        >>> my_feature(model.l2)
+        >>> model(torch.randn(2, 10)).sum().backward()
+    """
+
+    def get_all_state(module: nn.Module) -> Dict[Callable, Dict]:
+        d = module.__dict__.setdefault(state_key, {})
+        assert isinstance(d, dict), "Distributed composable API states corrupted"
+        return d
+
+    def wrapper(module: nn.Module, *args, **kwargs) -> Optional[nn.Module]:
+        # install states specific to the wrapped ``func``
+        all_state: Dict[Callable, dict] = get_all_state(module)
+        assert func not in all_state, (
+            "Each distinct composable distributed API can only be applied to a "
+            f"module once. {func} has already been applied to the following "
+            f"module.\n{module}"
+        )
+        all_state.setdefault(func, _State())
+
+        orig_named_params = OrderedDict(module.named_parameters())
+        orig_named_buffers = OrderedDict(
+            module.named_buffers(remove_duplicate=False)
```

Comment:
Seems `module.named_buffers()` has no arg of `remove_duplicate`:
https://pytorch.org/docs/stable/generated/torch.nn.Module.html#torch.nn.Module.named_buffers

Reply:
The code has the arg:

https://github.com/pytorch/pytorch/blob/1e1b04512879d6166dc1f5adff482723e2d0da9e/torch/nn/modules/module.py#L2011

Let me update `nn.Module`'s doc as well

Reply:
Oh, I see, it's added recently: https://github.com/pytorch/pytorch/pull/84984 :)